cvmfs/talk.cc

```diff
@@ -308,10 +311,56 @@ static void *MainTalk(void *data __attribute__((unused))) {
           if (proxies == "") {
               Answer(con_fd, "Failed, no valid proxies\n");
           } else {
-            cvmfs::download_manager_->SetProxyChain(proxies);
+            // get existing fallback proxies
+            vector< vector<download::DownloadManager::ProxyInfo> > proxy_chain;
+            unsigned fallback_group;
+            cvmfs::download_manager_->GetProxyInfo(&proxy_chain, NULL, &fallback_group);
+            std::string fallback_proxies;
+            for (unsigned i = fallback_group; i < proxy_chain.size(); ++i) {
+              if (fallback_proxies.size() != 0)
+                fallback_proxies.append(";");
+              if (proxy_chain[i].size() > 0)
+                // fallback proxies may have only one per group
+                fallback_proxies.append(proxy_chain[i][0].Print());
+            }
+            cvmfs::download_manager_->SetProxyChain(proxies, fallback_proxies);
             Answer(con_fd, "OK\n");
           }
         }
+      } else if (line.substr(0, 14) == "proxy fallback") {
+        if (line.length() < 15) {
+          Answer(con_fd, "Usage: proxy fallback <proxy list>\n");
+        } else {
+          string fallback_proxies = line.substr(15);
+          vector< vector<download::DownloadManager::ProxyInfo> > proxy_chain;
+          unsigned fallback_group;
+
+          // get existing non-fallback proxies
+          cvmfs::download_manager_->GetProxyInfo(&proxy_chain, NULL, &fallback_group);
+          std::string proxies;
+          for (unsigned i = 0; i < fallback_group; ++i) {
+            if (proxy_chain[i].size() == 0) {
```

Comment:
In this case there may not have been a fallback set earlier, so the original proxy list may still have a DIRECT.